src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,708 @@
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "(ref object x2) => { _ = x2.Length; }").WithLocation(8, 23),
+                // (8,51): error CS1061: 'object' does not contain a definition for 'Length' and no accessible extension method 'Length' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)
+                //         Delegate d2 = (ref object x2) => { _ = x2.Length; };
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "Length").WithArguments("object", "Length").WithLocation(8, 51),
+                // (9,53): error CS1061: 'object' does not contain a definition for 'Length' and no accessible extension method 'Length' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)
+                //         Delegate d3 = delegate (object x3) { _ = x3.Length; };
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "Length").WithArguments("object", "Length").WithLocation(9, 53));
+        }
+
+        [Fact]
+        public void InstanceMethods()
+        {
+            var source =
+@"using System;
+class Program
+{
+    object F1() => null;
+    void F2(object x, int y) { }
+    void F()
+    {
+        Delegate d1 = F1;
+        Delegate d2 = this.F2;
+        Console.WriteLine(""{0}, {1}"", d1.GetType().Name, d2.GetType().Name);
+    }
+    static void Main()
+    {
+        new Program().F();
+    }
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput: "Func`1, Action`2");
+        }
+
+        [Fact]
+        public void ExtensionMethods_01()
+        {
+            var source =
+@"static class E
+{
+    internal static void F1(this object x, int y) { }
+    internal static void F2(this object x) { }
+}
+class Program
+{
+    void F2(int x) { }
+    static void Main()
+    {
+        System.Delegate d;
+        var p = new Program();
+        d = p.F1;
+        d = p.F2;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (14,15): error CS8915: The delegate type could not be inferred.
+                //         d = p.F2;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F2").WithLocation(14, 15));
+        }
+
+        [Fact]
+        public void ExtensionMethods_02()
+        {
+            var source =
+@"static class E
+{
+    internal static void F(this System.Type x, int y) { }
+    internal static void F(this string x) { }
+}
+class Program
+{
+    static void Main()
+    {
+        System.Delegate d;
+        d = typeof(Program).F;
+        d = """".F;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+        }
```

Comment:
If we don't have tests for that already, can we check the natural type (on semantic model) in extension scenario?

Reply:
Tested type from the semantic model and at runtime. (Note, the semantic model currently does not expose the natural type.)